<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pbcore.io.dataset &mdash; pbcoretools 0.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/pacbio.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pbcoretools 0.1.0 documentation" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pbcoretools 0.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pbcore-io-dataset">
<h1>pbcore.io.dataset<a class="headerlink" href="#pbcore-io-dataset" title="Permalink to this headline">¶</a></h1>
<p>The Python DataSet XML API is designed to be a lightweight interface for
creating, opening, manipulating and writing DataSet XML files. It provides both
a native Python API and console entry points for use in manual dataset curation
or as a resource for P_Module developers.</p>
<p>The API and console entry points are designed with the set operations one might
perform on the various types of data held by a DataSet XML in mind: merge,
split, write etc. While various types of DataSets can be found in XML files,
the API (and in a way the console entry point, dataset.py) has DataSet as its
base type, with various subtypes extending or replacing functionality as
needed.</p>
</div>
<div class="section" id="console-entry-point-usage">
<h1>Console Entry Point Usage<a class="headerlink" href="#console-entry-point-usage" title="Permalink to this headline">¶</a></h1>
<p>The following entry points are available through the main script: dataset.py:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py [-h] [-v] [--debug]
                  {create,filter,merge,split,validate,loadstats,consolidate}
                  ...

Run dataset.py by specifying a command.

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program&#39;s version number and exit
  --debug               Turn on debug level logging

DataSet sub-commands:
  {create,filter,merge,split,validate,loadstats,consolidate}
                        Type {command} -h for a command&#39;s options
</pre></div>
</div>
<p>Create:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py create [-h] [--type DSTYPE] [--novalidate] [--relative]
                         outfile infile [infile ...]

Create an XML file from a fofn or bam

positional arguments:
  outfile        The XML to create
  infile         The fofn or BAM file(s) to make into an XML

optional arguments:
  -h, --help     show this help message and exit
  --type DSTYPE  The type of XML to create
  --novalidate   Don&#39;t validate the resulting XML, don&#39;t touch paths
  --relative     Make the included paths relative instead of absolute (not
                 compatible with --novalidate)
</pre></div>
</div>
<p>Filter:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py filter [-h] infile outfile filters [filters ...]

Add filters to an XML file. Suggested fields: [&#39;bcf&#39;, &#39;bcq&#39;, &#39;bcr&#39;,
&#39;length&#39;, &#39;pos&#39;, &#39;qend&#39;, &#39;qname&#39;, &#39;qstart&#39;, &#39;readstart&#39;, &#39;rname&#39;, &#39;rq&#39;,
&#39;tend&#39;, &#39;tstart&#39;, &#39;zm&#39;]. More expensive fields: [&#39;accuracy&#39;, &#39;bc&#39;, &#39;movie&#39;,
&#39;qs&#39;]

positional arguments:
  infile      The xml file to filter
  outfile     The resulting xml file
  filters     The values and thresholds to filter (e.g. &#39;rq&gt;0.85&#39;)

optional arguments:
  -h, --help  show this help message and exit
</pre></div>
</div>
<p>Union:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py union [-h] outfile infiles [infiles ...]

Combine XML (and BAM) files

positional arguments:
  outfile     The resulting XML file
  infiles     The XML files to merge

optional arguments:
  -h, --help  show this help message and exit
</pre></div>
</div>
<p>Validate:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py validate [-h] infile

Validate ResourceId files (XML validation only available in testing)

positional arguments:
  infile      The XML file to validate

optional arguments:
  -h, --help  show this help message and exit
</pre></div>
</div>
<p>Load PipeStats:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py loadstats [-h] [--outfile OUTFILE] infile statsfile

Load an sts.xml file into a DataSet XML file

positional arguments:
  infile             The XML file to modify
  statsfile          The .sts.xml file to load

optional arguments:
  -h, --help         show this help message and exit
  --outfile OUTFILE  The XML file to output
</pre></div>
</div>
<p>Split:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py split [-h] [--contigs] [--chunks CHUNKS] [--subdatasets]
                        [--outdir OUTDIR]
                        infile ...

Split the dataset

positional arguments:
  infile           The xml file to split
  outfiles         The resulting xml files

optional arguments:
  -h, --help       show this help message and exit
  --contigs        Split on contigs
  --chunks CHUNKS  Split contigs into &lt;chunks&gt; total windows
  --subdatasets    Split on subdatasets
  --outdir OUTDIR  Specify an output directory
</pre></div>
</div>
<p>Consolidate:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: dataset.py consolidate [-h] [--numFiles NUMFILES] [--noTmp]
                              infile datafile xmlfile

Consolidate the XML files

positional arguments:
  infile               The XML file to consolidate
  datafile             The resulting data file
  xmlfile              The resulting XML file

optional arguments:
  -h, --help           show this help message and exit
  --numFiles NUMFILES  The number of data files to produce (1)
  --noTmp              Don&#39;t copy to a tmp location to ensure local disk
                       use
</pre></div>
</div>
</div>
<div class="section" id="usage-examples">
<h1>Usage Examples<a class="headerlink" href="#usage-examples" title="Permalink to this headline">¶</a></h1>
<div class="section" id="filter-reads-cli-version">
<h2>Filter Reads (CLI version)<a class="headerlink" href="#filter-reads-cli-version" title="Permalink to this headline">¶</a></h2>
<p>In this scenario we have one or more bam files worth of subreads, aligned or
otherwise, that we want to filter and put in a single bam file. This is
possible using the CLI with the following steps, starting with a DataSet XML
file:</p>
<div class="highlight-python"><div class="highlight"><pre># usage: dataset.py filter &lt;in_fn.xml&gt; &lt;out_fn.xml&gt; &lt;filters&gt;
dataset.py filter in_fn.subreadset.xml filtered_fn.subreadset.xml &#39;rq&gt;0.85&#39;

# usage: dataset.py consolidate &lt;in_fn.xml&gt; &lt;out_data_fn.bam&gt; &lt;out_fn.xml&gt;
dataset.py consolidate filtered_fn.subreadset.xml consolidate.subreads.bam out_fn.subreadset.xml
</pre></div>
</div>
<p>The filtered DataSet and the consolidated DataSet should be read for read
equivalent when used with SMRT Analysis software.</p>
</div>
<div class="section" id="filter-reads-api-version">
<h2>Filter Reads (API version)<a class="headerlink" href="#filter-reads-api-version" title="Permalink to this headline">¶</a></h2>
<p>The API version of filtering allows for more advanced filtering criteria:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ss</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="s">&#39;in_fn.subreadset.xml&#39;</span><span class="p">)</span>
<span class="n">ss</span><span class="o">.</span><span class="n">filters</span><span class="o">.</span><span class="n">addRequirement</span><span class="p">(</span><span class="n">rname</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="s">&#39;E.faecalis.2&#39;</span><span class="p">),</span>
                                 <span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="s">&#39;E.faecalis.2&#39;</span><span class="p">)],</span>
                          <span class="n">tStart</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;&lt;&#39;</span><span class="p">,</span> <span class="s">&#39;99&#39;</span><span class="p">),</span>
                                  <span class="p">(</span><span class="s">&#39;&lt;&#39;</span><span class="p">,</span> <span class="s">&#39;299&#39;</span><span class="p">)],</span>
                          <span class="n">tEnd</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s">&#39;0&#39;</span><span class="p">),</span>
                                <span class="p">(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s">&#39;200&#39;</span><span class="p">)])</span>
</pre></div>
</div>
<p>Produces the following conditions for a read to be considered passing:</p>
<p>(rname = E.faecalis.2 AND tstart &lt; 99 AND tend &gt; 0)
OR
(rname = E.faecalis.2 AND tstart &lt; 299 AND tend &gt; 200)</p>
<p>You can add sets of filters by providing equal length lists of requirements for
each filter.</p>
<p>Additional requirements added singly will be added to all filters:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ss</span><span class="o">.</span><span class="n">filters</span><span class="o">.</span><span class="n">addRequirement</span><span class="p">(</span><span class="n">rq</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s">&#39;0.85&#39;</span><span class="p">)])</span>
</pre></div>
</div>
<p>(rname = E.faecalis.2 AND tstart &lt; 99 AND tend &gt; 0 AND rq &gt; 0.85)
OR
(rname = E.faecalis.2 AND tstart &lt; 299 AND tend &gt; 100 AND rq &gt; 0.85)</p>
<p>Additional requirements added with a plurality of options will duplicate the
previous requiremnts for each option:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ss</span><span class="o">.</span><span class="n">filters</span><span class="o">.</span><span class="n">addRequirement</span><span class="p">(</span><span class="n">length</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">),</span> <span class="p">(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)])</span>
</pre></div>
</div>
<p>(rname = E.faecalis.2 AND tstart &lt; 99 AND tend &gt; 0 AND rq &gt; 0.85 AND length &gt; 500)
OR
(rname = E.faecalis.2 AND tstart &lt; 299 AND tend &gt; 100 AND rq &gt; 0.85 AND length &gt; 500)
OR
(rname = E.faecalis.2 AND tstart &lt; 99 AND tend &gt; 0 AND rq &gt; 0.85 AND length &gt; 1000)
OR
(rname = E.faecalis.2 AND tstart &lt; 299 AND tend &gt; 100 AND rq &gt; 0.85 AND length &gt; 1000)</p>
<p>Of course you can always wipe the filters and start over:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ss</span><span class="o">.</span><span class="n">filters</span> <span class="o">=</span> <span class="bp">None</span>
</pre></div>
</div>
<p>Consolidation is more similar to the CLI version:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ss</span><span class="o">.</span><span class="n">consolidate</span><span class="p">(</span><span class="s">&#39;cons.bam&#39;</span><span class="p">)</span>
<span class="n">ss</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;cons.xml&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="resequencing-pipeline-cli-version">
<h2>Resequencing Pipeline (CLI version)<a class="headerlink" href="#resequencing-pipeline-cli-version" title="Permalink to this headline">¶</a></h2>
<p>In this scenario, we have two movies worth of subreads in two SubreadSets that
we want to align to a reference, merge together, split into DataSet chunks by
contig, then send through quiver on a chunkwise basis (in parallel).</p>
<ol class="arabic">
<li><p class="first">Align each movie to the reference, producing a dataset with one bam file for
each execution:</p>
<div class="highlight-python"><div class="highlight"><pre>pbalign movie1.subreadset.xml referenceset.xml movie1.alignmentset.xml
pbalign movie2.subreadset.xml referenceset.xml movie2.alignmentset.xml
</pre></div>
</div>
</li>
<li><p class="first">Merge the files into a FOFN-like dataset (bams aren&#8217;t touched):</p>
<div class="highlight-python"><div class="highlight"><pre># dataset.py merge &lt;out_fn&gt; &lt;in_fn&gt; [&lt;in_fn&gt; &lt;in_fn&gt; ...]
dataset.py merge merged.alignmentset.xml movie1.alignmentset.xml movie2.alignmentset.xml
</pre></div>
</div>
</li>
<li><p class="first">Split the dataset into chunks by contig (rname) (bams aren&#8217;t touched). Note
that supplying output files splits the dataset into that many output files
(up to the number of contigs), with multiple contigs per file. Not supplying
output files splits the dataset into one output file per contig, named
automatically. Specifying a number of chunks instead will produce that many
files, with contig or even sub contig (reference window) splitting.:</p>
<div class="highlight-python"><div class="highlight"><pre>dataset.py split --contigs --chunks 8 merged.alignmentset.xml
</pre></div>
</div>
</li>
<li><p class="first">Quiver then consumes these chunks:</p>
<div class="highlight-python"><div class="highlight"><pre>variantCaller.py --alignmentSetRefWindows --referenceFileName referenceset.xml --outputFilename chunk1consensus.fasta --algorithm quiver chunk1contigs.alignmentset.xml
variantCaller.py --alignmentSetRefWindows --referenceFileName referenceset.xml --outputFilename chunk2consensus.fasta --algorithm quiver chunk2contigs.alignmentset.xml
</pre></div>
</div>
</li>
</ol>
<p>The chunking works by duplicating the original merged dataset (no bam
duplication) and adding filters to each duplicate such that only reads
belonging to the appropriate contigs are emitted. The contigs are distributed
amongst the output files in such a way that the total number of records per
chunk is about even.</p>
<div class="section" id="tangential-information">
<h3>Tangential Information<a class="headerlink" href="#tangential-information" title="Permalink to this headline">¶</a></h3>
<p>DataSet.refNames (which returns a list of reference names available in the
dataset) is also subject to the filtering imposed during the split. Therefore
you wont be running through superfluous (and apparently unsampled) contigs to
get the reads in this chunk. The DataSet.records generator is also subject to
filtering, but not as efficiently as readsInRange. If you do not have a
reference window, readsInReference() is also an option.</p>
<p>As the bam files are never touched, each dataset contains all the information
necessary to access all reads for all contigs. Doing so on these filtered
datasets would require disabling the filters first:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dset</span><span class="o">.</span><span class="n">disableFilters</span><span class="p">()</span>
</pre></div>
</div>
<p>Or removing the specific filter giving you problems:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dset</span><span class="o">.</span><span class="n">filters</span><span class="o">.</span><span class="n">removeRequirement</span><span class="p">(</span><span class="s">&#39;rname&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="resequencing-pipeline-api-version">
<h2>Resequencing Pipeline (API version)<a class="headerlink" href="#resequencing-pipeline-api-version" title="Permalink to this headline">¶</a></h2>
<p>In this scenario, we have two movies worth of subreads in two SubreadSets that
we want to align to a reference, merge together, split into DataSet chunks by
contig, then send through quiver on a chunkwise basis (in parallel). We want to
do them using the API, rather than the CLI.</p>
<ol class="arabic">
<li><p class="first">Align each movie to the reference, producing a dataset with one bam file for
each execution</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># CLI (or see pbalign API):</span>
<span class="n">pbalign</span> <span class="n">movie1</span><span class="o">.</span><span class="n">subreadset</span><span class="o">.</span><span class="n">xml</span> <span class="n">referenceset</span><span class="o">.</span><span class="n">xml</span> <span class="n">movie1</span><span class="o">.</span><span class="n">alignmentset</span><span class="o">.</span><span class="n">xml</span>
<span class="n">pbalign</span> <span class="n">movie2</span><span class="o">.</span><span class="n">subreadset</span><span class="o">.</span><span class="n">xml</span> <span class="n">referenceset</span><span class="o">.</span><span class="n">xml</span> <span class="n">movie2</span><span class="o">.</span><span class="n">alignmentset</span><span class="o">.</span><span class="n">xml</span>
</pre></div>
</div>
</li>
<li><p class="first">Merge the files into a FOFN-like dataset (bams aren&#8217;t touched)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># API, filename_list is dummy data:</span>
<span class="n">filename_list</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;movie1.alignmentset.xml&#39;</span><span class="p">,</span> <span class="s">&#39;movie2.alignmentset.xml&#39;</span><span class="p">]</span>

<span class="c"># open:</span>
<span class="n">dsets</span> <span class="o">=</span> <span class="p">[</span><span class="n">AlignmentSet</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span> <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">filename_list</span><span class="p">]</span>
<span class="c"># merge with + operator:</span>
<span class="n">dset</span> <span class="o">=</span> <span class="nb">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">y</span><span class="p">,</span> <span class="n">dsets</span><span class="p">)</span>

<span class="c"># OR:</span>
<span class="n">dset</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="o">*</span><span class="n">filename_list</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">Split the dataset into chunks by contigs (or subcontig windows)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># split:</span>
<span class="n">dsets</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">contigs</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chunks</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">Quiver then consumes these chunks</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># write out if you need to (or pass directly to quiver API):</span>
<span class="n">outfilename_list</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;chunk1contigs.alignmentset.xml&#39;</span><span class="p">,</span> <span class="s">&#39;chunk2contigs.alignmentset.xml&#39;</span><span class="p">]</span>
<span class="c"># write with &#39;write&#39; method:</span>
<span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">nm</span><span class="p">):</span> <span class="n">ds</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">nm</span><span class="p">),</span> <span class="nb">zip</span><span class="p">(</span><span class="n">dsets</span><span class="p">,</span> <span class="n">outfilename_list</span><span class="p">))</span>

<span class="c"># CLI (or see quiver API):</span>
<span class="n">variantCaller</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">alignmentSetRefWindows</span> <span class="o">--</span><span class="n">referenceFileName</span> <span class="n">referenceset</span><span class="o">.</span><span class="n">xml</span> <span class="o">--</span><span class="n">outputFilename</span> <span class="n">chunk1consensus</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">algorithm</span> <span class="n">quiver</span> <span class="n">chunk1contigs</span><span class="o">.</span><span class="n">alignmentset</span><span class="o">.</span><span class="n">xml</span>
<span class="n">variantCaller</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">alignmentSetRefWindows</span> <span class="o">--</span><span class="n">referenceFileName</span> <span class="n">referenceset</span><span class="o">.</span><span class="n">xml</span> <span class="o">--</span><span class="n">outputFilename</span> <span class="n">chunk2consensus</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">algorithm</span> <span class="n">quiver</span> <span class="n">chunk2contigs</span><span class="o">.</span><span class="n">alignmentset</span><span class="o">.</span><span class="n">xml</span>

<span class="c"># Inside quiver (still using python dataset API):</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
<span class="c"># get this set&#39;s windows:</span>
<span class="n">refWindows</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">refWindows</span>
<span class="c"># gather the reads for these windows using readsInRange, e.g.:</span>
<span class="n">reads</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">readsInRange</span><span class="p">(</span><span class="n">rId</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span> <span class="k">for</span> <span class="n">rId</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">refWindows</span><span class="p">))</span>
</pre></div>
</div>
</li>
</ol>
</div>
</div>
<div class="section" id="api-overview">
<h1>API overview<a class="headerlink" href="#api-overview" title="Permalink to this headline">¶</a></h1>
<p>The chunking works by duplicating the original merged dataset (no bam
duplication) and adding filters to each duplicate such that only reads
belonging to the appropriate contigs/windows are emitted. The contigs are
distributed amongst the output files in such a way that the total number of
records per chunk is about even.</p>
<p>DataSets can be created using the appropriate constructor (SubreadSet), or with
the common constructor (DataSet) and later cast to a specific type
(copy(asType=&#8221;SubreadSet&#8221;)). The DataSet constructor acts as a factory function
(an artifact of early api Designs). The factory behavior is defined in the
DataSet metaclass.</p>
<p class="graphviz">
<img src="_images/inheritance-a18070efaf9136370707ea0b6a7998b0b887a456.png" alt="Inheritance diagram of pbcore.io.dataset.DataSetIO" usemap="#inheritance5d781fce17" class="inheritance"/>
<map id="inheritance5d781fce17" name="inheritance5d781fce17">
<area shape="rect" id="node1" href="#pbcore.io.dataset.DataSetIO.AlignmentSet" title="DataSet type specific to Alignments. No type specific Metadata exists," alt="" coords="428,6,531,30"/>
<area shape="rect" id="node5" href="#pbcore.io.dataset.DataSetIO.ConsensusAlignmentSet" title="Dataset type for aligned CCS reads. &#160;Essentially identical to AlignmentSet" alt="" coords="594,6,763,30"/>
<area shape="rect" id="node2" href="#pbcore.io.dataset.DataSetIO.ReadSet" title="Base type for read sets, should probably never be used as a concrete" alt="" coords="248,99,319,123"/>
<area shape="rect" id="node6" href="#pbcore.io.dataset.DataSetIO.ConsensusReadSet" title="DataSet type specific to CCSreads. No type specific Metadata exists, so" alt="" coords="411,52,548,76"/>
<area shape="rect" id="node13" href="#pbcore.io.dataset.DataSetIO.SubreadSet" title="DataSet type specific to Subreads" alt="" coords="433,145,526,169"/>
<area shape="rect" id="node3" href="#pbcore.io.dataset.DataSetIO.BarcodeSet" title="DataSet type specific to Barcodes" alt="" coords="237,145,329,169"/>
<area shape="rect" id="node4" href="#pbcore.io.dataset.DataSetIO.DataSet" title="The record containing the DataSet information, with possible type" alt="" coords="45,145,116,169"/>
<area shape="rect" id="node7" href="#pbcore.io.dataset.DataSetIO.ContigSet" title="DataSet type specific to Contigs" alt="" coords="243,192,324,216"/>
<area shape="rect" id="node11" href="#pbcore.io.dataset.DataSetIO.ReferenceSet" title="DataSet type specific to References" alt="" coords="429,192,530,216"/>
<area shape="rect" id="node8" href="#pbcore.io.dataset.DataSetIO.DataSetMetaTypes" title="This mirrors the PacBioSecondaryDataModel.xsd definitions and be used" alt="" coords="13,99,148,123"/>
<area shape="rect" id="node10" href="#pbcore.io.dataset.DataSetIO.InvalidDataSetIOError" title="The base class for all DataSetIO related custom exceptions" alt="" coords="6,52,155,76"/>
</map>
</p>
<span class="target" id="module-pbcore.io.dataset.DataSetIO"></span><p>Classes representing DataSets of various types.</p>
<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">AlignmentSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.ReadSet" title="pbcore.io.dataset.DataSetIO.ReadSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.ReadSet</span></code></a></p>
<p>DataSet type specific to Alignments. No type specific Metadata exists,
so the base class version is OK (this just ensures type representation on
output and expandability</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>An AlignmentSet</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">files:</th><td class="field-body">handled by super</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">referenceFastaFname=None:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">the reference fasta filename for this                                        alignment.</td>
</tr>
<tr class="field-odd field"><th class="field-name">strict=False:</th><td class="field-body">see base class</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">skipCounts=False:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">see base class</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.countRecords">
<code class="descname">countRecords</code><span class="sig-paren">(</span><em>rname=None</em>, <em>winStart=None</em>, <em>winEnd=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.countRecords" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the number of records mapped to &#8216;rname&#8217; that overlap with
&#8216;window&#8217;</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.fullRefNames">
<code class="descname">fullRefNames</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.fullRefNames" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of reference full names (full header).</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.readsInRange">
<code class="descname">readsInRange</code><span class="sig-paren">(</span><em>refName</em>, <em>start</em>, <em>end</em>, <em>buffsize=50</em>, <em>usePbi=True</em>, <em>longest=False</em>, <em>sampleSize=0</em>, <em>justIndices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.readsInRange" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of (usually) BamAlignment objects for the reads in one
or more Bam files pointed to by the ExternalResources in this DataSet
that have at least one coordinate within the specified range in the
reference genome.</p>
<p>Rather than developing some convoluted approach for dealing with
auto-inferring the desired references, this method and self.refNames
should allow users to compose the desired query.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">refName:</th><td class="field-body">the name of the reference that we are sampling</td>
</tr>
<tr class="field-even field"><th class="field-name">start:</th><td class="field-body">the start of the range (inclusive, index relative to                     reference)</td>
</tr>
<tr class="field-odd field"><th class="field-name">end:</th><td class="field-body">the end of the range (inclusive, index relative to reference)</td>
</tr>
</tbody>
</table>
</dd>
<dt>Yields:</dt>
<dd>BamAlignment objects</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">readsInRange</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">refNames</span><span class="p">[</span><span class="mi">15</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&#39;hn: </span><span class="si">%i</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">read</span><span class="o">.</span><span class="n">holeNumber</span> 
<span class="go">hn: ...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.readsInReference">
<code class="descname">readsInReference</code><span class="sig-paren">(</span><em>refName</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.readsInReference" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of (usually) BamAlignment objects for the
reads in one or more Bam files pointed to by the ExternalResources in
this DataSet that are mapped to the specified reference genome.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">refName:</th><td class="field-body">the name of the reference that we are sampling.</td>
</tr>
</tbody>
</table>
</dd>
<dt>Yields:</dt>
<dd>BamAlignment objects</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">readsInReference</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">refNames</span><span class="p">[</span><span class="mi">15</span><span class="p">]):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&#39;hn: </span><span class="si">%i</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">read</span><span class="o">.</span><span class="n">holeNumber</span> 
<span class="go">hn: ...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.records">
<code class="descname">records</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.records" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of (usually) BamAlignment objects for the
records in one or more Bam files pointed to by the
ExternalResources in this DataSet.</p>
<dl class="docutils">
<dt>Yields:</dt>
<dd>A BamAlignment object</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&#39;hn: </span><span class="si">%i</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">holeNumber</span> 
<span class="go">hn: ...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.recordsByReference">
<code class="descname">recordsByReference</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.recordsByReference" title="Permalink to this definition">¶</a></dt>
<dd><p>The records in this AlignmentSet, sorted by tStart.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refIds">
<code class="descname">refIds</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refIds" title="Permalink to this definition">¶</a></dt>
<dd><p>A dict of refName: refId for the joined referenceInfoTable</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refInfo">
<code class="descname">refInfo</code><span class="sig-paren">(</span><em>key</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a column in the referenceInfoTable, tupled with the reference
name. TODO(mdsmith)(2016-01-27): pick a better name for this method...</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refLength">
<code class="descname">refLength</code><span class="sig-paren">(</span><em>rname</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refLength" title="Permalink to this definition">¶</a></dt>
<dd><p>The length of reference &#8216;rname&#8217;. This is expensive, so if you&#8217;re
going to do many lookups cache self.refLengths locally and use that.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refLengths">
<code class="descname">refLengths</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refLengths" title="Permalink to this definition">¶</a></dt>
<dd><p>A dict of refName: refLength</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refNames">
<code class="descname">refNames</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refNames" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of reference names (id).</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.refWindows">
<code class="descname">refWindows</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.refWindows" title="Permalink to this definition">¶</a></dt>
<dd><p>Going to be tricky unless the filters are really focused on
windowing the reference. Much nesting or duplication and the correct
results are really not guaranteed</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.referenceInfo">
<code class="descname">referenceInfo</code><span class="sig-paren">(</span><em>refName</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.referenceInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Select a row from the DataSet.referenceInfoTable using the reference
name as a unique key (or ID, if you really have to)</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.referenceInfoTable">
<code class="descname">referenceInfoTable</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.referenceInfoTable" title="Permalink to this definition">¶</a></dt>
<dd><p>The merged reference info tables from the external resources.
Record.ID is remapped to a unique integer key (though using record.Name
is preferred). Record.Names are remapped for cmp.h5 files to be
consistent with bam files.</p>
<p>..note:: Reference names are assumed to be unique</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.AlignmentSet.resourceReaders">
<code class="descname">resourceReaders</code><span class="sig-paren">(</span><em>refName=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.AlignmentSet.resourceReaders" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of Indexed*Reader objects for the ExternalResources
in this DataSet.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">refName:</th><td class="field-body">Only yield open resources if they have refName in their
referenceInfoTable</td>
</tr>
</tbody>
</table>
</dd>
<dt>Yields:</dt>
<dd>An open indexed alignment file</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seqFile</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">resourceReaders</span><span class="p">():</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">seqFile</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="s">&#39;hn: </span><span class="si">%i</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">holeNumber</span> 
<span class="go">hn: ...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.BarcodeSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">BarcodeSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.BarcodeSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.DataSet" title="pbcore.io.dataset.DataSetIO.DataSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.DataSet</span></code></a></p>
<p>DataSet type specific to Barcodes</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.BarcodeSet.addMetadata">
<code class="descname">addMetadata</code><span class="sig-paren">(</span><em>newMetadata</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.BarcodeSet.addMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Add metadata specific to this subtype, while leaning on the
superclass method for generic metadata. Also enforce metadata type
correctness.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.ConsensusAlignmentSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">ConsensusAlignmentSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ConsensusAlignmentSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.AlignmentSet" title="pbcore.io.dataset.DataSetIO.AlignmentSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.AlignmentSet</span></code></a></p>
<p>Dataset type for aligned CCS reads.  Essentially identical to AlignmentSet
aside from the contents of the underlying BAM files.</p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.ConsensusReadSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">ConsensusReadSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ConsensusReadSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.ReadSet" title="pbcore.io.dataset.DataSetIO.ReadSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.ReadSet</span></code></a></p>
<p>DataSet type specific to CCSreads. No type specific Metadata exists, so
the base class version is OK (this just ensures type representation on
output and expandability</p>
<dl class="docutils">
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">ConsensusReadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">ConsensusReadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">strict</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> 
<span class="go">&lt;ConsensusReadSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">ContigSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.DataSet" title="pbcore.io.dataset.DataSetIO.DataSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.DataSet</span></code></a></p>
<p>DataSet type specific to Contigs</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.addMetadata">
<code class="descname">addMetadata</code><span class="sig-paren">(</span><em>newMetadata</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.addMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Add metadata specific to this subtype, while leaning on the
superclass method for generic metadata. Also enforce metadata type
correctness.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.consolidate">
<code class="descname">consolidate</code><span class="sig-paren">(</span><em>outfn=None</em>, <em>numFiles=1</em>, <em>useTmp=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.consolidate" title="Permalink to this definition">¶</a></dt>
<dd><p>Consolidation should be implemented for window text in names and
for filters in ContigSets</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.contigNames">
<code class="descname">contigNames</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.contigNames" title="Permalink to this definition">¶</a></dt>
<dd><p>The names assigned to the External Resources, or contigs if no name
assigned.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.contigs">
<code class="descname">contigs</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.contigs" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of contigs from the fastaReader objects for the
ExternalResources in this ReferenceSet.</p>
<dl class="docutils">
<dt>Yields:</dt>
<dd>A fasta file entry</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.get_contig">
<code class="descname">get_contig</code><span class="sig-paren">(</span><em>contig_id</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.get_contig" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a contig by ID</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ContigSet.resourceReaders">
<code class="descname">resourceReaders</code><span class="sig-paren">(</span><em>refName=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ContigSet.resourceReaders" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of fastaReader objects for the ExternalResources in this
ReferenceSet.</p>
<dl class="docutils">
<dt>Yields:</dt>
<dd>An open fasta file</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.DataSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">DataSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>The record containing the DataSet information, with possible type
specific subclasses</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.__add__">
<code class="descname">__add__</code><span class="sig-paren">(</span><em>otherDataset</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge the representations of two DataSets without modifying
the original datasets. (Fails if filters are incompatible).</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">otherDataset:</th><td class="field-body">a DataSet to merge with self</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>A new DataSet with members containing the union of the input
DataSets&#8217; members and subdatasets representing the input DataSets</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io.dataset.DataSetWriter</span> <span class="kn">import</span> <span class="n">toXml</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># xmls with different resourceIds: success</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">11</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span> <span class="o">+</span> <span class="n">ds2</span><span class="o">.</span><span class="n">numExternalResources</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span><span class="o">.</span><span class="n">numExternalResources</span> <span class="o">==</span> <span class="n">expected</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># xmls with different resourceIds but conflicting filters:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># failure to merge</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">filters</span><span class="o">.</span><span class="n">addRequirement</span><span class="p">(</span><span class="n">rname</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="s">&#39;E.faecalis.1&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># xmls with same resourceIds: ignores new inputs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span><span class="o">.</span><span class="n">numExternalResources</span> <span class="o">==</span> <span class="n">expected</span>
<span class="go">True</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.__deepcopy__">
<code class="descname">__deepcopy__</code><span class="sig-paren">(</span><em>memo</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.__deepcopy__" title="Permalink to this definition">¶</a></dt>
<dd><p>Deep copy this Dataset by recursively deep copying the members
(objMetadata, DataSet metadata, externalResources, filters and
subdatasets)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.__eq__">
<code class="descname">__eq__</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Test for DataSet equality. The method specified in the documentation
calls for md5 hashing the &#8220;Core XML&#8221; elements and comparing. This is
the same procedure for generating the Uuid, so the same method may be
used. However, as simultaneously or regularly updating the Uuid is not
specified, we opt to not set the newUuid when checking for equality.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">other:</th><td class="field-body">The other DataSet to compare to this DataSet.</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>T/F the Core XML elements of this and the other DataSet hash to the
same value</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>DataSet constructor</p>
<p>Initialize representations of the ExternalResources, MetaData,
Filters, and LabeledSubsets, parse inputs if possible</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">files:</th><td class="field-body">one or more filenames or uris to read</td>
</tr>
<tr class="field-even field"><th class="field-name">strict=False:</th><td class="field-body">strictly require all index files</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">skipCounts=False:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">skip updating counts for faster opening</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">tempfile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span><span class="p">,</span> <span class="n">SubreadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Prog like pbalign provides a .bam file:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># e.g. d = AlignmentSet(&quot;aligned.bam&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Something like the test files we have:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inBam</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">inBam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># A UniqueId is generated, despite being a BAM input</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">uuid</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dOldUuid</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">uuid</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># They can write this BAM to an XML:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># e.g. d.write(&quot;alignmentset.xml&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s">&quot;dataset-doctest&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outXml</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s">&#39;tempfile.xml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outXml</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># And then recover the same XML:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">outXml</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># The UniqueId will be the same</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">uuid</span> <span class="o">==</span> <span class="n">dOldUuid</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Inputs can be many and varied</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">8</span><span class="p">),</span> <span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getFofn</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Constructors should be used directly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getSubreadSet</span><span class="p">())</span> 
<span class="go">&lt;SubreadSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Even with untyped inputs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span> 
<span class="go">&lt;AlignmentSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># AlignmentSets can also be manipulated after opening:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Add external Resources:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="o">.</span><span class="n">addResources</span><span class="p">([</span><span class="s">&quot;IdontExist.bam&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">resourceId</span> <span class="o">==</span> <span class="s">&quot;IdontExist.bam&quot;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Add an index file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pbiName</span> <span class="o">=</span> <span class="s">&quot;IdontExist.bam.pbi&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">addIndices</span><span class="p">([</span><span class="n">pbiName</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">resourceId</span> <span class="o">==</span> <span class="n">pbiName</span>
<span class="go">True</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.__repr__">
<code class="descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the dataset with an informative string:</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd>&#8220;&lt;type uuid filenames&gt;&#8221;</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.addDatasets">
<code class="descname">addDatasets</code><span class="sig-paren">(</span><em>otherDataSet</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.addDatasets" title="Permalink to this definition">¶</a></dt>
<dd><p>Add subsets to a DataSet object using other DataSets.</p>
<p>The following method of enabling merge-based split prevents nesting of
datasets more than one deep. Nested relationships are flattened.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Most often used by the __add__ method, rather than directly.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.addExternalResources">
<code class="descname">addExternalResources</code><span class="sig-paren">(</span><em>newExtResources</em>, <em>updateCount=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.addExternalResources" title="Permalink to this definition">¶</a></dt>
<dd><p>Add additional ExternalResource objects, ensuring no duplicate
resourceIds. Most often used by the __add__ method, rather than
directly.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name" colspan="2">newExtResources:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">A list of new ExternalResource objects, either
created de novo from a raw bam input, parsed from
an xml input, or already contained in a separate
DataSet object and being merged.</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io.dataset.DataSetMembers</span> <span class="kn">import</span> <span class="n">ExternalResource</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># it is possible to add ExtRes&#39;s as ExternalResource objects:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er1</span> <span class="o">=</span> <span class="n">ExternalResource</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er1</span><span class="o">.</span><span class="n">resourceId</span> <span class="o">=</span> <span class="s">&quot;test1.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er2</span> <span class="o">=</span> <span class="n">ExternalResource</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er2</span><span class="o">.</span><span class="n">resourceId</span> <span class="o">=</span> <span class="s">&quot;test2.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er3</span> <span class="o">=</span> <span class="n">ExternalResource</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">er3</span><span class="o">.</span><span class="n">resourceId</span> <span class="o">=</span> <span class="s">&quot;test1.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">addExternalResources</span><span class="p">([</span><span class="n">er1</span><span class="p">],</span> <span class="n">updateCount</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># different resourceId: succeeds</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">addExternalResources</span><span class="p">([</span><span class="n">er2</span><span class="p">],</span> <span class="n">updateCount</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># same resourceId: fails</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">addExternalResources</span><span class="p">([</span><span class="n">er3</span><span class="p">],</span> <span class="n">updateCount</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># but it is probably better to add them a little deeper:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">externalResources</span><span class="o">.</span><span class="n">addResources</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[</span><span class="s">&quot;test3.bam&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">addIndices</span><span class="p">([</span><span class="s">&quot;test3.bam.bai&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.addFilters">
<code class="descname">addFilters</code><span class="sig-paren">(</span><em>newFilters</em>, <em>underConstruction=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.addFilters" title="Permalink to this definition">¶</a></dt>
<dd><p>Add new or extend the current list of filters. Public because there
is already a reasonably compelling reason (the console script entry
point). Most often used by the __add__ method.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">newFilters:</th><td class="field-body">a Filters object or properly formatted Filters record</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io.dataset.DataSetMembers</span> <span class="kn">import</span> <span class="n">Filters</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">Filters</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">addRequirement</span><span class="p">(</span><span class="n">rq</span><span class="o">=</span><span class="p">[(</span><span class="s">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s">&#39;0.85&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">addFilters</span><span class="p">(</span><span class="n">filt</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">ds1</span><span class="o">.</span><span class="n">filters</span>
<span class="go">( rq &gt; 0.85 )</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Or load with a DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">16</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">ds2</span><span class="o">.</span><span class="n">filters</span>
<span class="gp">... </span>
<span class="go">( rname = E.faecalis...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.addMetadata">
<code class="descname">addMetadata</code><span class="sig-paren">(</span><em>newMetadata</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.addMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Add dataset metadata.</p>
<p>Currently we ignore duplicates while merging (though perhaps other
transformations are more appropriate) and plan to remove/merge
conflicting metadata with identical attribute names.</p>
<p>All metadata elements should be strings, deepcopy shouldn&#8217;t be
necessary.</p>
<p>This method is most often used by the __add__ method, rather than
directly.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">newMetadata:</th><td class="field-body">a dictionary of object metadata from an XML file (or
carefully crafted to resemble one), or a wrapper
around said dictionary</td>
</tr>
<tr class="field-even field"><th class="field-name">kwargs:</th><td class="field-body">new metadata fields to be piled into the current metadata
(as an attribute)</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># it is possible to add new metadata:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">addMetadata</span><span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="n">Name</span><span class="o">=</span><span class="s">&#39;LongReadsRock&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">ds</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">getV</span><span class="p">(</span><span class="n">container</span><span class="o">=</span><span class="s">&#39;attrib&#39;</span><span class="p">,</span> <span class="n">tag</span><span class="o">=</span><span class="s">&#39;Name&#39;</span><span class="p">)</span>
<span class="go">LongReadsRock</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># but most will be loaded and modified:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">totalLength</span>
<span class="go">123588</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">totalLength</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">totalLength</span>
<span class="go">100000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">totalLength</span> <span class="o">+=</span> <span class="mi">100000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">_metadata</span><span class="o">.</span><span class="n">totalLength</span>
<span class="go">200000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span><span class="o">.</span><span class="n">loadStats</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getStats</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds4</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">11</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds4</span><span class="o">.</span><span class="n">loadStats</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getStats</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds5</span> <span class="o">=</span> <span class="n">ds3</span> <span class="o">+</span> <span class="n">ds4</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.barcodes">
<code class="descname">barcodes</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.barcodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the list of barcodes explicitly set by filters via
DataSet.split(barcodes=True).</p>
</dd></dl>

<dl class="classmethod">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.castableTypes">
<em class="property">classmethod </em><code class="descname">castableTypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.castableTypes" title="Permalink to this definition">¶</a></dt>
<dd><p>The types to which this DataSet type may be cast. This is a property
instead of a member variable as we can enforce casting limits here (and
modify if needed by overriding them in subclasses).</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd>A dictionary of MetaType-&gt;Class mappings, e.g. &#8216;DataSet&#8217;: DataSet</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.close">
<code class="descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close all of the opened resource readers</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><em>asType=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deep copy the representation of this DataSet</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">asType:</th><td class="field-body">The type of DataSet to return, e.g. &#8216;AlignmentSet&#8217;</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>A DataSet object that is identical but for UniqueId</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span><span class="p">,</span> <span class="n">SubreadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">12</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Deep copying datasets is easy:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># But the resulting uuid&#39;s should be different.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">==</span> <span class="n">ds2</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">uuid</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">uuid</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="ow">is</span> <span class="n">ds2</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Most members are identical</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">name</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">externalResources</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">externalResources</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">filters</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">filters</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">subdatasets</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">subdatasets</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds1</span><span class="o">.</span><span class="n">subdatasets</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">subdatasets</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Except for the one that stores the uuid:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">objMetadata</span> <span class="o">==</span> <span class="n">ds2</span><span class="o">.</span><span class="n">objMetadata</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># And of course identical != the same object:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="ow">not</span> <span class="nb">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="ow">or</span> <span class="n">y</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="p">[</span><span class="n">ds1d</span> <span class="ow">is</span> <span class="n">ds2d</span> <span class="k">for</span> <span class="n">ds1d</span> <span class="ow">in</span>
<span class="gp">... </span>                   <span class="n">ds1</span><span class="o">.</span><span class="n">subdatasets</span> <span class="k">for</span> <span class="n">ds2d</span> <span class="ow">in</span>
<span class="gp">... </span>                   <span class="n">ds2</span><span class="o">.</span><span class="n">subdatasets</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># But types are maintained:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span> <span class="n">strict</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Lets try casting</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> 
<span class="go">&lt;DataSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">asType</span><span class="o">=</span><span class="s">&#39;SubreadSet&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> 
<span class="go">&lt;SubreadSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Lets do some illicit casting</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">asType</span><span class="o">=</span><span class="s">&#39;ReferenceSet&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="gr">TypeError</span>: <span class="n">Cannot cast from SubreadSet to ReferenceSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Lets try not having to cast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> 
<span class="go">&lt;SubreadSet...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.copyFiles">
<code class="descname">copyFiles</code><span class="sig-paren">(</span><em>outdir</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.copyFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy all of the top level ExternalResources to an output
directory &#8216;outdir&#8217;</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.createdAt">
<code class="descname">createdAt</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.createdAt" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the DataSet CreatedAt timestamp</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.description">
<code class="descname">description</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.description" title="Permalink to this definition">¶</a></dt>
<dd><p>The description of this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.disableFilters">
<code class="descname">disableFilters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.disableFilters" title="Permalink to this definition">¶</a></dt>
<dd><p>Disable read filtering for this object</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.enableFilters">
<code class="descname">enableFilters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.enableFilters" title="Permalink to this definition">¶</a></dt>
<dd><p>Re-enable read filtering for this object</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.filters">
<code class="descname">filters</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Limit setting to ensure cache hygiene and filter compatibility</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.induceIndices">
<code class="descname">induceIndices</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.induceIndices" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate indices for ExternalResources.</p>
<p>Not compatible with DataSet base type</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.loadStats">
<code class="descname">loadStats</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.loadStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Load pipeline statistics from a &lt;moviename&gt;.sts.xml file. The subset
of these data that are defined in the DataSet XSD become available
through via DataSet.metadata.summaryStats.&lt;...&gt; and will be written out
to the DataSet XML format according to the DataSet XML XSD.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">filename:</th><td class="field-body">the filename of a &lt;moviename&gt;.sts.xml file</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">loadStats</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getStats</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">11</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">loadStats</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getStats</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">summaryStats</span><span class="o">.</span><span class="n">prodDist</span><span class="o">.</span><span class="n">bins</span>
<span class="go">[1576, 901, 399, 0]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">summaryStats</span><span class="o">.</span><span class="n">prodDist</span><span class="o">.</span><span class="n">bins</span>
<span class="go">[1576, 901, 399, 0]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">summaryStats</span><span class="o">.</span><span class="n">prodDist</span><span class="o">.</span><span class="n">bins</span>
<span class="go">[3152, 1802, 798, 0]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.makePathsAbsolute">
<code class="descname">makePathsAbsolute</code><span class="sig-paren">(</span><em>curStart='.'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.makePathsAbsolute" title="Permalink to this definition">¶</a></dt>
<dd><p>As part of the validation process, make all ResourceIds absolute
URIs rather than relative paths. Generally not called by API users.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">curStart:</th><td class="field-body">The location from which relative paths should emanate.</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.makePathsRelative">
<code class="descname">makePathsRelative</code><span class="sig-paren">(</span><em>outDir=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.makePathsRelative" title="Permalink to this definition">¶</a></dt>
<dd><p>Make things easier for writing test cases: make all
ResourceIds relative paths rather than absolute paths.
A less common use case for API consumers.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">outDir:</th><td class="field-body">The location from which relative paths should originate</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em>, <em>copyOnMerge=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge an &#8216;other&#8217; dataset with this dataset, same as add operator,
but can take argumens</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.metadata">
<code class="descname">metadata</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the DataSet metadata as a DataSetMetadata object. Attributes
should be populated intuitively, but see DataSetMetadata documentation
for more detail.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.name">
<code class="descname">name</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.name" title="Permalink to this definition">¶</a></dt>
<dd><p>The name of this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.newUuid">
<code class="descname">newUuid</code><span class="sig-paren">(</span><em>setter=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.newUuid" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and enforce the uniqueness of an ID for a new DataSet.
While user setable fields are stripped out of the Core DataSet object
used for comparison, the previous UniqueId is not. That means that
copies will still be unique, despite having the same contents.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">setter=True:</th><td class="field-body">Setting to False allows MD5 hashes to be generated
(e.g. for comparison with other objects) without
modifying the object&#8217;s UniqueId</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>The new Id, a properly formatted md5 hash of the Core DataSet</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">old</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">uuid</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">newUuid</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">old</span> <span class="o">!=</span> <span class="n">ds</span><span class="o">.</span><span class="n">uuid</span>
<span class="go">True</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.numExternalResources">
<code class="descname">numExternalResources</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.numExternalResources" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of ExternalResources in this DataSet</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.numRecords">
<code class="descname">numRecords</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.numRecords" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of records in this DataSet (from the metadata)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.processFilters">
<code class="descname">processFilters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.processFilters" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a list of functions to apply to a read, all of which return
T/F. Each function is an OR filter, so any() true passes the read.
These functions are the AND filters, and will likely check all() of
other functions. These filtration functions are cached so that they are
not regenerated from the base filters for every read</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.reFilter">
<code class="descname">reFilter</code><span class="sig-paren">(</span><em>light=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.reFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>The filters on this dataset have changed, update DataSet state as
needed</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.readsInSubDatasets">
<code class="descname">readsInSubDatasets</code><span class="sig-paren">(</span><em>subNames=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.readsInSubDatasets" title="Permalink to this definition">¶</a></dt>
<dd><p>To be used in conjunction with self.subSetNames</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.records">
<code class="descname">records</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.records" title="Permalink to this definition">¶</a></dt>
<dd><p>A generator of (usually) BamAlignment objects for the
records in one or more Bam files pointed to by the
ExternalResources in this DataSet.</p>
<dl class="docutils">
<dt>Yields:</dt>
<dd>A BamAlignment object</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getBam</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&#39;hn: </span><span class="si">%i</span><span class="s">&#39;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">holeNumber</span> 
<span class="go">hn: ...</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.resourceReaders">
<code class="descname">resourceReaders</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.resourceReaders" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of open pbcore Reader objects for the
top level ExternalResources in this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.split">
<code class="descname">split</code><span class="sig-paren">(</span><em>chunks=0</em>, <em>ignoreSubDatasets=True</em>, <em>contigs=False</em>, <em>maxChunks=0</em>, <em>breakContigs=False</em>, <em>targetSize=5000</em>, <em>zmws=False</em>, <em>barcodes=False</em>, <em>byRecords=False</em>, <em>updateCounts=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Deep copy the DataSet into a number of new DataSets containing
roughly equal chunks of the ExternalResources or subdatasets.</p>
<p>Examples:</p>
<blockquote>
<div><ul>
<li><p class="first">split into exactly n datasets where each addresses a different               piece of the collection of contigs:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">contigs</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chunks</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">split into at most n datasets where each addresses a different               piece of the collection of contigs, but contigs are kept whole:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">contigs</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">maxChunks</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">split into at most n datasets where each addresses a different               piece of the collection of contigs and the number of chunks is               in part based on the number of reads:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dset</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">contigs</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">maxChunks</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">breakContigs</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
</div></blockquote>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">chunks:</th><td class="field-body">the number of chunks to split the DataSet.</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">ignoreSubDatasets:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">(True) do not split on subdatasets</td>
</tr>
<tr class="field-odd field"><th class="field-name">contigs:</th><td class="field-body">split on contigs instead of external resources etc</td>
</tr>
<tr class="field-even field"><th class="field-name">maxChunks:</th><td class="field-body">The upper limit on the number of chunks.</td>
</tr>
<tr class="field-odd field"><th class="field-name">breakContigs:</th><td class="field-body">Whether or not to break contigs when using maxChunks</td>
</tr>
<tr class="field-even field"><th class="field-name">targetSize:</th><td class="field-body">The target number of reads per chunk</td>
</tr>
<tr class="field-odd field"><th class="field-name">zmws:</th><td class="field-body">Split by zmws</td>
</tr>
<tr class="field-even field"><th class="field-name">barcodes:</th><td class="field-body">Split by barcodes</td>
</tr>
<tr class="field-odd field"><th class="field-name">byRecords:</th><td class="field-body">Split contigs by mapped records, rather than ref length</td>
</tr>
<tr class="field-even field"><th class="field-name">updateCounts:</th><td class="field-body">Update the count metadata in each chunk</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>A list of new DataSet objects (all other information deep copied).</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">AlignmentSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># splitting is pretty intuitive:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">12</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># but divides up extRes&#39;s, so have more than one:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span> <span class="o">&gt;</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># the default is one AlignmentSet per ExtRes:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">dss</span><span class="p">)</span> <span class="o">==</span> <span class="n">ds1</span><span class="o">.</span><span class="n">numExternalResources</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># but you can specify a number of AlignmentSets to produce:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">chunks</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">dss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">chunks</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignoreSubDatasets</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">dss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># The resulting objects are similar:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">uuid</span> <span class="o">==</span> <span class="n">dss</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">uuid</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">dss</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Previously merged datasets are &#39;unmerged&#39; upon split, unless</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># otherwise specified.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Lets try merging and splitting on subdatasets:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">totalLength</span>
<span class="go">123588</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1tl</span> <span class="o">=</span> <span class="n">ds1</span><span class="o">.</span><span class="n">totalLength</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">AlignmentSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">11</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">totalLength</span>
<span class="go">117086</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2tl</span> <span class="o">=</span> <span class="n">ds2</span><span class="o">.</span><span class="n">totalLength</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># merge:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dss</span><span class="o">.</span><span class="n">totalLength</span> <span class="o">==</span> <span class="p">(</span><span class="n">ds1tl</span> <span class="o">+</span> <span class="n">ds2tl</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># unmerge:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="p">,</span> <span class="n">ds2</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
<span class="gp">... </span><span class="n">dss</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignoreSubDatasets</span><span class="o">=</span><span class="bp">False</span><span class="p">),</span>
<span class="gp">... </span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">totalLength</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">totalLength</span> <span class="o">==</span> <span class="n">ds1tl</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">totalLength</span> <span class="o">==</span> <span class="n">ds2tl</span>
<span class="go">True</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.subSetNames">
<code class="descname">subSetNames</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.subSetNames" title="Permalink to this definition">¶</a></dt>
<dd><p>The subdataset names present in this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.toExternalFiles">
<code class="descname">toExternalFiles</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.toExternalFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of top level external resources (no indices).</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.toFofn">
<code class="descname">toFofn</code><span class="sig-paren">(</span><em>outfn=None</em>, <em>uri=False</em>, <em>relative=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.toFofn" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of resource filenames (and write to optional outfile)</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">outfn:</th><td class="field-body">(None) the file to which the resouce filenames are to be
written. If None, the only emission is a returned list of
file names.</td>
</tr>
<tr class="field-even field"><th class="field-name">uri:</th><td class="field-body">(t/F) write the resource filenames as URIs.</td>
</tr>
<tr class="field-odd field"><th class="field-name">relative:</th><td class="field-body">(t/F) emit paths relative to outfofn or &#8216;.&#8217; if no
outfofn</td>
</tr>
</tbody>
</table>
</dd>
<dt>Returns:</dt>
<dd>A list of filenames or uris</dd>
<dt>Writes:</dt>
<dd>(Optional) A file containing a list of filenames or uris</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">DataSet</span><span class="p">(</span><span class="s">&quot;bam1.bam&quot;</span><span class="p">,</span> <span class="s">&quot;bam2.bam&quot;</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span><span class="o">.</span><span class="n">toFofn</span><span class="p">(</span><span class="n">uri</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">[&#39;bam1.bam&#39;, &#39;bam2.bam&#39;]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.totalLength">
<code class="descname">totalLength</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.totalLength" title="Permalink to this definition">¶</a></dt>
<dd><p>The total length of this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.updateCounts">
<code class="descname">updateCounts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.updateCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the TotalLength and NumRecords for this DataSet.</p>
<p>Not compatible with the base DataSet class, which has no ability to
touch ExternalResources. -1 is used as a sentinel value for failed size
determination. It should never be written out to XML in regular use.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.uuid">
<code class="descname">uuid</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.uuid" title="Permalink to this definition">¶</a></dt>
<dd><p>The UniqueId of this DataSet</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.write">
<code class="descname">write</code><span class="sig-paren">(</span><em>outFile</em>, <em>validate=True</em>, <em>modPaths=False</em>, <em>relPaths=False</em>, <em>pretty=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write to disk as an XML file</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">outFile:</th><td class="field-body">The filename of the xml file to be created</td>
</tr>
<tr class="field-even field"><th class="field-name">validate:</th><td class="field-body">T/F (True) validate the ExternalResource ResourceIds</td>
</tr>
<tr class="field-odd field"><th class="field-name">relPaths:</th><td class="field-body">T/F (False) make the ExternalResource ResourceIds
relative instead of absolute filenames</td>
</tr>
</tbody>
</table>
</dd>
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tempfile</span><span class="o">,</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s">&quot;dataset-doctest&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s">&#39;tempfile.xml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">outfile</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">==</span> <span class="n">ds2</span>
<span class="go">True</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.DataSet.zmwRanges">
<code class="descname">zmwRanges</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSet.zmwRanges" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the end-inclusive range of ZMWs covered by the dataset if this
was explicitly set by filters via DataSet.split(zmws=True).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.DataSetMetaTypes">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">DataSetMetaTypes</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.DataSetMetaTypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>This mirrors the PacBioSecondaryDataModel.xsd definitions and be used
to reference a specific dataset type.</p>
</dd></dl>

<dl class="exception">
<dt id="pbcore.io.dataset.DataSetIO.InvalidDataSetIOError">
<em class="property">exception </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">InvalidDataSetIOError</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.InvalidDataSetIOError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></a></p>
<p>The base class for all DataSetIO related custom exceptions</p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">ReadSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.DataSet" title="pbcore.io.dataset.DataSetIO.DataSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.DataSet</span></code></a></p>
<p>Base type for read sets, should probably never be used as a concrete
class</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.addMetadata">
<code class="descname">addMetadata</code><span class="sig-paren">(</span><em>newMetadata</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.addMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Add metadata specific to this subtype, while leaning on the
superclass method for generic metadata. Also enforce metadata type
correctness.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.assertBarcoded">
<code class="descname">assertBarcoded</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.assertBarcoded" title="Permalink to this definition">¶</a></dt>
<dd><p>Test whether all resources are barcoded files</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.consolidate">
<code class="descname">consolidate</code><span class="sig-paren">(</span><em>dataFile</em>, <em>numFiles=1</em>, <em>useTmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.consolidate" title="Permalink to this definition">¶</a></dt>
<dd><p>Consolidate a larger number of bam files to a smaller number of bam
files (min 1)</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><table class="first last docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">dataFile:</th><td class="field-body">The name of the output file. If numFiles &gt;1 numbers will
be added.</td>
</tr>
<tr class="field-even field"><th class="field-name">numFiles:</th><td class="field-body">The number of data files to be produced.</td>
</tr>
</tbody>
</table>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.hasPbi">
<code class="descname">hasPbi</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.hasPbi" title="Permalink to this definition">¶</a></dt>
<dd><p>Test whether all resources are opened as IndexedBamReader objects</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.isCmpH5">
<code class="descname">isCmpH5</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.isCmpH5" title="Permalink to this definition">¶</a></dt>
<dd><p>Test whether all resources are cmp.h5 files</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.movieIds">
<code class="descname">movieIds</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.movieIds" title="Permalink to this definition">¶</a></dt>
<dd><p>A dict of movieName: movieId for the joined readGroupTable</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.readGroupTable">
<code class="descname">readGroupTable</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.readGroupTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine the readGroupTables of each external resource</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetIO.ReadSet.resourceReaders">
<code class="descname">resourceReaders</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReadSet.resourceReaders" title="Permalink to this definition">¶</a></dt>
<dd><p>Open the files in this ReadSet</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.ReferenceSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">ReferenceSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReferenceSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.ContigSet" title="pbcore.io.dataset.DataSetIO.ContigSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.ContigSet</span></code></a></p>
<p>DataSet type specific to References</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetIO.ReferenceSet.refNames">
<code class="descname">refNames</code><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.ReferenceSet.refNames" title="Permalink to this definition">¶</a></dt>
<dd><p>The reference names assigned to the External Resources, or contigs
if no name assigned.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetIO.SubreadSet">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">SubreadSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.SubreadSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetIO.ReadSet" title="pbcore.io.dataset.DataSetIO.ReadSet"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetIO.ReadSet</span></code></a></p>
<p>DataSet type specific to Subreads</p>
<p>DocTest:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">SubreadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io.dataset.DataSetMembers</span> <span class="kn">import</span> <span class="n">ExternalResources</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="n">no</span><span class="o">=</span><span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># So they don&#39;t conflict:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">externalResources</span> <span class="o">=</span> <span class="n">ExternalResources</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> 
<span class="go">&lt;SubreadSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">_metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">_metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds3</span> <span class="o">=</span> <span class="n">ds1</span> <span class="o">+</span> <span class="n">ds2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds3</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds4</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getSubreadSet</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds4</span> 
<span class="go">&lt;SubreadSet...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds4</span><span class="o">.</span><span class="n">_metadata</span> 
<span class="go">&lt;SubreadSetMetadata...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ds4</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetIO.checkAndResolve">
<code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">checkAndResolve</code><span class="sig-paren">(</span><em>fname</em>, <em>possibleRelStart=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.checkAndResolve" title="Permalink to this definition">¶</a></dt>
<dd><p>Try and skip resolveLocation if possible</p>
</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetIO.filtered">
<code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">filtered</code><span class="sig-paren">(</span><em>generator</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.filtered" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrap a generator with postfiltering</p>
</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetIO.isDataSet">
<code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">isDataSet</code><span class="sig-paren">(</span><em>xmlfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.isDataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine if a file is a DataSet before opening it</p>
</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetIO.openDataFile">
<code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">openDataFile</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.openDataFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory function for DataSet types determined by the first data file</p>
</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetIO.openDataSet">
<code class="descclassname">pbcore.io.dataset.DataSetIO.</code><code class="descname">openDataSet</code><span class="sig-paren">(</span><em>*files</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetIO.openDataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory function for DataSet types as suggested by the first file</p>
</dd></dl>

<p>The operations possible between DataSets of the same and different types are
limited, see the DataSet XML documentation for details.</p>
<p>DataSet XML files have a few major components: XML metadata,
ExternalReferences, Filters, DataSet Metadata, etc. These are represented in
different ways internally, depending on their complexity. DataSet metadata
especially contains a large number of different potential elements, many of
which are accessible in the API as nested attributes. To preserve the API&#8217;s
ability to grant access to any DataSet Metadata available now and in the
future, as well as to maintain the performance of dataset reading and writing,
each DataSet stores its metadata in what approximates a tree structure, with
various helper classes and functions manipulating this tree. The structure of
this tree and currently implemented helper classes are available in the
DataSetMembers module.</p>
<p class="graphviz">
<img src="_images/inheritance-a39314a9102e499cc8dc7919537c3b5f2e8b349a.png" alt="Inheritance diagram of pbcore.io.dataset.DataSetMembers" usemap="#inheritance1742c48be3" class="inheritance"/>
<map id="inheritance1742c48be3" name="inheritance1742c48be3">
<area shape="rect" id="node1" href="#pbcore.io.dataset.DataSetMembers.Automation" title="Automation" alt="" coords="179,113,251,132"/>
<area shape="rect" id="node2" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="The base functionality of a metadata element." alt="" coords="11,622,97,640"/>
<area shape="rect" id="node3" href="#pbcore.io.dataset.DataSetMembers.AutomationParameter" title="AutomationParameter" alt="" coords="155,150,275,168"/>
<area shape="rect" id="node4" href="#pbcore.io.dataset.DataSetMembers.AutomationParameters" title="AutomationParameters" alt="" coords="152,186,278,205"/>
<area shape="rect" id="node6" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata" title="The root of the DataSetMetadata element tree, used as base for subtype" alt="" coords="165,222,265,241"/>
<area shape="rect" id="node11" href="#pbcore.io.dataset.DataSetMembers.BioSampleMetadata" title="The metadata for a single BioSample" alt="" coords="159,259,271,277"/>
<area shape="rect" id="node12" href="#pbcore.io.dataset.DataSetMembers.BioSamplePointersMetadata" title="The BioSamplePointer members don&#39;t seem complex enough to justify" alt="" coords="140,295,290,314"/>
<area shape="rect" id="node13" href="#pbcore.io.dataset.DataSetMembers.BioSamplesMetadata" title="The metadata for the list of BioSamples" alt="" coords="157,331,273,350"/>
<area shape="rect" id="node14" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata" title="The metadata for a single collection. It contains Context," alt="" coords="161,367,268,386"/>
<area shape="rect" id="node15" href="#pbcore.io.dataset.DataSetMembers.CollectionsMetadata" title="The Element should just have children: a list of" alt="" coords="159,404,271,422"/>
<area shape="rect" id="node16" href="#pbcore.io.dataset.DataSetMembers.ContigMetadata" title="ContigMetadata" alt="" coords="169,440,261,459"/>
<area shape="rect" id="node18" href="#pbcore.io.dataset.DataSetMembers.ContigsMetadata" title="ContigsMetadata" alt="" coords="167,476,263,495"/>
<area shape="rect" id="node19" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution" title="ContinuousDistribution" alt="" coords="154,513,276,531"/>
<area shape="rect" id="node20" href="#pbcore.io.dataset.DataSetMembers.CopyFilesMetadata" title="The CopyFile members don&#39;t seem complex enough to justify" alt="" coords="162,549,267,568"/>
<area shape="rect" id="node21" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution" title="DiscreteDistribution" alt="" coords="160,585,269,604"/>
<area shape="rect" id="node22" href="#pbcore.io.dataset.DataSetMembers.ExternalResource" title="ExternalResource" alt="" coords="166,622,263,640"/>
<area shape="rect" id="node23" href="#pbcore.io.dataset.DataSetMembers.ExternalResources" title="ExternalResources" alt="" coords="164,658,266,677"/>
<area shape="rect" id="node24" href="#pbcore.io.dataset.DataSetMembers.FileIndex" title="FileIndex" alt="" coords="187,694,243,713"/>
<area shape="rect" id="node25" href="#pbcore.io.dataset.DataSetMembers.FileIndices" title="FileIndices" alt="" coords="183,731,247,749"/>
<area shape="rect" id="node26" href="#pbcore.io.dataset.DataSetMembers.Filter" title="Filter" alt="" coords="188,767,241,785"/>
<area shape="rect" id="node27" href="#pbcore.io.dataset.DataSetMembers.Filters" title="Filters" alt="" coords="188,803,241,822"/>
<area shape="rect" id="node28" href="#pbcore.io.dataset.DataSetMembers.OutputOptions" title="OutputOptions" alt="" coords="173,839,257,858"/>
<area shape="rect" id="node29" href="#pbcore.io.dataset.DataSetMembers.ParentTool" title="ParentTool" alt="" coords="183,876,247,894"/>
<area shape="rect" id="node31" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata" title="Doctest:" alt="" coords="166,912,263,931"/>
<area shape="rect" id="node32" href="#pbcore.io.dataset.DataSetMembers.Properties" title="Properties" alt="" coords="183,948,247,967"/>
<area shape="rect" id="node33" href="#pbcore.io.dataset.DataSetMembers.Property" title="Property" alt="" coords="187,985,243,1003"/>
<area shape="rect" id="node34" href="#pbcore.io.dataset.DataSetMembers.Provenance" title="The metadata concerning this dataset&#39;s provenance" alt="" coords="179,1021,251,1040"/>
<area shape="rect" id="node35" href="#pbcore.io.dataset.DataSetMembers.RunDetailsMetadata" title="RunDetailsMetadata" alt="" coords="159,1057,270,1076"/>
<area shape="rect" id="node36" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata" title="The metadata from the machine sts.xml" alt="" coords="172,1094,258,1112"/>
<area shape="rect" id="node38" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata" title="WellSampleMetadata" alt="" coords="157,1130,272,1149"/>
<area shape="rect" id="node5" href="#pbcore.io.dataset.DataSetMembers.BarcodeSetMetadata" title="The DataSetMetadata subtype specific to BarcodeSets." alt="" coords="326,186,443,205"/>
<area shape="rect" id="node17" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata" title="The DataSetMetadata subtype specific to ContigSets." alt="" coords="331,222,439,241"/>
<area shape="rect" id="node37" href="#pbcore.io.dataset.DataSetMembers.SubreadSetMetadata" title="The DataSetMetadata subtype specific to SubreadSets. Deals explicitly" alt="" coords="326,259,443,277"/>
<area shape="rect" id="node7" href="#pbcore.io.dataset.DataSetMembers.BinBoundaryMismatchError" title="BinBoundaryMismatchError" alt="" coords="143,4,287,23"/>
<area shape="rect" id="node8" href="#pbcore.io.dataset.DataSetMembers.BinMismatchError" title="BinMismatchError" alt="" coords="4,41,104,59"/>
<area shape="rect" id="node9" href="#pbcore.io.dataset.DataSetMembers.BinNumberMismatchError" title="BinNumberMismatchError" alt="" coords="146,41,284,59"/>
<area shape="rect" id="node10" href="#pbcore.io.dataset.DataSetMembers.BinWidthMismatchError" title="BinWidthMismatchError" alt="" coords="152,77,278,96"/>
<area shape="rect" id="node30" href="#pbcore.io.dataset.DataSetMembers.PbiFlags" title="PbiFlags" alt="" coords="26,4,81,23"/>
</map>
</p>
<span class="target" id="module-pbcore.io.dataset.DataSetMembers"></span><p>DataSetMetadata (also the tag of the Element in the DataSet XML
representation) is somewhat challening to store, access, and (de)serialize
efficiently. Here, we maintain a bulk representation of all of the dataset
metadata (or any other XML data, like ExternalResources) found in the XML file
in the following data structure:</p>
<div class="line-block">
<div class="line">An Element is a turned into a dictionary:</div>
<div class="line-block">
<div class="line">XmlElement =&gt; {&#8216;tag&#8217;: &#8216;ElementTag&#8217;,</div>
<div class="line-block">
<div class="line">&#8216;text&#8217;: &#8216;ElementText&#8217;,</div>
<div class="line">&#8216;attrib&#8217;: {&#8216;ElementAttributeName&#8217;: &#8216;AttributeValue&#8217;,
&#8216;AnotherAttrName&#8217;: &#8216;AnotherAttrValue&#8217;},</div>
<div class="line">&#8216;children&#8217;: [XmlElementDict,
XmlElementDictWithSameOrDifferentTag]}</div>
</div>
</div>
</div>
<p>Child elements are represented similarly and stored (recursively) as a list
in &#8216;children&#8217;. The top level we store for DataSetMetadata is just a list,
which can be thought of as the list of children of a different element
(say, a DataSet or SubreadSet element, if we stored that):</p>
<ul class="simple">
<li>DataSetMetadata = [XmlTag, XmlTagWithSameOrDifferentTag]</li>
</ul>
<p>We keep this for three reasons:</p>
<ol class="arabic simple">
<li>We don&#8217;t want to have to write a lot of logic to go from XML to an internal representation and then back to XML.</li>
<li>We want to be able to store and at least write metadata that doesn&#8217;t yet exist, even if we can&#8217;t merge it intelligently.</li>
<li>Keeping and manipulating a dictionary is ~10x faster than an OrderedAttrDict, and probably faster to use than a full stack of objects.</li>
</ol>
<p>Instead, we keep and modify this list:dictionary structure, wrapping it in
classes as necessary. The classes and methods that wrap this datastructure
serve two pruposes:</p>
<ul>
<li><dl class="first docutils">
<dt>Provide an interface for our code (and making merging clean) e.g.:</dt>
<dd><ul class="first last simple">
<li>DataSet(&#8220;test.xml&#8221;).metadata.numRecords += 1</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Provide an interface for users of the DataSet API, e.g.:</dt>
<dd><ul class="first last">
<li><p class="first">numRecords = DataSet(&#8220;test.xml&#8221;).metadata.numRecords</p>
</li>
<li><p class="first">bioSamplePointer = (DataSet(&#8220;test.xml&#8221;)                          .metadata.collections[0]                          .wellSample.bioSamplePointers[0])</p>
</li>
<li><dl class="first docutils">
<dt>Though users can still access novel metadata types the hard way e.g.:</dt>
<dd><ul class="first last simple">
<li>bioSamplePointer = (DataSet(&#8220;test.xml&#8221;)                              .metadata.collections[0]                              [&#8216;WellSample&#8217;][&#8216;BioSamplePointers&#8217;]                              [&#8216;BioSamplePointer&#8217;].record[&#8216;text&#8217;])</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Automation">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Automation</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Automation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Automation.automationParameters">
<code class="descname">automationParameters</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Automation.automationParameters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.AutomationParameter">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">AutomationParameter</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.AutomationParameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.AutomationParameters">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">AutomationParameters</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.AutomationParameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.AutomationParameters.addParameter">
<code class="descname">addParameter</code><span class="sig-paren">(</span><em>key</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.AutomationParameters.addParameter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.AutomationParameters.automationParameter">
<code class="descname">automationParameter</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.AutomationParameters.automationParameter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.BarcodeSetMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BarcodeSetMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BarcodeSetMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata" title="pbcore.io.dataset.DataSetMembers.DataSetMetadata"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.DataSetMetadata</span></code></a></p>
<p>The DataSetMetadata subtype specific to BarcodeSets.</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.BarcodeSetMetadata.barcodeConstruction">
<code class="descname">barcodeConstruction</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BarcodeSetMetadata.barcodeConstruction" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="exception">
<dt id="pbcore.io.dataset.DataSetMembers.BinBoundaryMismatchError">
<em class="property">exception </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BinBoundaryMismatchError</code><span class="sig-paren">(</span><em>min1</em>, <em>min2</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BinBoundaryMismatchError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.BinMismatchError" title="pbcore.io.dataset.DataSetMembers.BinMismatchError"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.BinMismatchError</span></code></a></p>
</dd></dl>

<dl class="exception">
<dt id="pbcore.io.dataset.DataSetMembers.BinMismatchError">
<em class="property">exception </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BinMismatchError</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BinMismatchError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></a></p>
</dd></dl>

<dl class="exception">
<dt id="pbcore.io.dataset.DataSetMembers.BinNumberMismatchError">
<em class="property">exception </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BinNumberMismatchError</code><span class="sig-paren">(</span><em>num1</em>, <em>num2</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BinNumberMismatchError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.BinMismatchError" title="pbcore.io.dataset.DataSetMembers.BinMismatchError"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.BinMismatchError</span></code></a></p>
</dd></dl>

<dl class="exception">
<dt id="pbcore.io.dataset.DataSetMembers.BinWidthMismatchError">
<em class="property">exception </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BinWidthMismatchError</code><span class="sig-paren">(</span><em>width1</em>, <em>width2</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BinWidthMismatchError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.BinMismatchError" title="pbcore.io.dataset.DataSetMembers.BinMismatchError"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.BinMismatchError</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.BioSampleMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BioSampleMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSampleMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The metadata for a single BioSample</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.BioSampleMetadata.createdAt">
<code class="descname">createdAt</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSampleMetadata.createdAt" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.BioSampleMetadata.uniqueId">
<code class="descname">uniqueId</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSampleMetadata.uniqueId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.BioSamplePointersMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BioSamplePointersMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSamplePointersMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The BioSamplePointer members don&#8217;t seem complex enough to justify
class representation, instead rely on base class methods to provide
iterators and accessors</p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.BioSamplesMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">BioSamplesMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSamplesMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The metadata for the list of BioSamples</p>
<dl class="docutils">
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">SubreadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getSubreadSet</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">bioSamples</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;consectetur purus&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">bs</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">bioSamples</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">bs</span><span class="o">.</span><span class="n">name</span>
<span class="go">consectetur purus</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">em</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;tag&#39;</span><span class="p">:</span><span class="s">&#39;BioSample&#39;</span><span class="p">,</span> <span class="s">&#39;text&#39;</span><span class="p">:</span><span class="s">&#39;&#39;</span><span class="p">,</span> <span class="s">&#39;children&#39;</span><span class="p">:[],</span>
<span class="gp">... </span>      <span class="s">&#39;attrib&#39;</span><span class="p">:{</span><span class="s">&#39;Name&#39;</span><span class="p">:</span><span class="s">&#39;great biosample&#39;</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">bioSamples</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">em</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">bioSamples</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;great biosample&#39;</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a biosample</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over biosamples</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.BioSamplesMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">CollectionMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The metadata for a single collection. It contains Context,
InstrumentName etc. as attribs, InstCtrlVer etc. for children</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.automation">
<code class="descname">automation</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.automation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.cellIndex">
<code class="descname">cellIndex</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.cellIndex" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.cellPac">
<code class="descname">cellPac</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.cellPac" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.collectionNumber">
<code class="descname">collectionNumber</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.collectionNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.context">
<code class="descname">context</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.context" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.instCtrlVer">
<code class="descname">instCtrlVer</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.instCtrlVer" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.instrumentId">
<code class="descname">instrumentId</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.instrumentId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.instrumentName">
<code class="descname">instrumentName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.instrumentName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.primary">
<code class="descname">primary</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.primary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.runDetails">
<code class="descname">runDetails</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.runDetails" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.sigProcVer">
<code class="descname">sigProcVer</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.sigProcVer" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionMetadata.wellSample">
<code class="descname">wellSample</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionMetadata.wellSample" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionsMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">CollectionsMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionsMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The Element should just have children: a list of
CollectionMetadataTags</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionsMetadata.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionsMetadata.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionsMetadata.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionsMetadata.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.CollectionsMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CollectionsMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ContigMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ContigMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContigMetadata.digest">
<code class="descname">digest</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigMetadata.digest" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContigMetadata.length">
<code class="descname">length</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigMetadata.length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ContigSetMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata" title="pbcore.io.dataset.DataSetMembers.DataSetMetadata"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.DataSetMetadata</span></code></a></p>
<p>The DataSetMetadata subtype specific to ContigSets.</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata.addContig">
<code class="descname">addContig</code><span class="sig-paren">(</span><em>newContig</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata.addContig" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata.contigs">
<code class="descname">contigs</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata.contigs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata.organism">
<code class="descname">organism</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata.organism" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContigSetMetadata.ploidy">
<code class="descname">ploidy</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigSetMetadata.ploidy" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ContigsMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ContigsMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigsMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContigsMetadata.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigsMetadata.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContigsMetadata.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigsMetadata.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContigsMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContigsMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ContinuousDistribution</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.binWidth">
<code class="descname">binWidth</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.binWidth" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.bins">
<code class="descname">bins</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.bins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.description">
<code class="descname">description</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.description" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.labels">
<code class="descname">labels</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Label the bins with the min value of each bin</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.maxBinValue">
<code class="descname">maxBinValue</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.maxBinValue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.maxOutlierValue">
<code class="descname">maxOutlierValue</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.maxOutlierValue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.minBinValue">
<code class="descname">minBinValue</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.minBinValue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.minOutlierValue">
<code class="descname">minOutlierValue</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.minOutlierValue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.numBins">
<code class="descname">numBins</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.numBins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sample95thPct">
<code class="descname">sample95thPct</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sample95thPct" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleMean">
<code class="descname">sampleMean</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleMean" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleMed">
<code class="descname">sampleMed</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleMed" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleSize">
<code class="descname">sampleSize</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleSize" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleStd">
<code class="descname">sampleStd</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ContinuousDistribution.sampleStd" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.CopyFilesMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">CopyFilesMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.CopyFilesMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The CopyFile members don&#8217;t seem complex enough to justify
class representation, instead rely on base class methods</p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">DataSetMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The root of the DataSetMetadata element tree, used as base for subtype
specific DataSet or for generic &#8220;DataSet&#8221; records.</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata.numRecords">
<code class="descname">numRecords</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata.numRecords" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of records in a DataSet using helper functions
defined in the base class</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata.provenance">
<code class="descname">provenance</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata.provenance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata.summaryStats">
<code class="descname">summaryStats</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata.summaryStats" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DataSetMetadata.totalLength">
<code class="descname">totalLength</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata.totalLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the TotalLength property of this dataset.
TODO: update the value from the actual external reference on
ValueError</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">DiscreteDistribution</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution.bins">
<code class="descname">bins</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution.bins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution.description">
<code class="descname">description</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution.description" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution.labels">
<code class="descname">labels</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution.labels" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.DiscreteDistribution.numBins">
<code class="descname">numBins</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.DiscreteDistribution.numBins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ExternalResource</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.addIndices">
<code class="descname">addIndices</code><span class="sig-paren">(</span><em>indices</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.addIndices" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.bai">
<code class="descname">bai</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.bai" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.bam">
<code class="descname">bam</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.bam" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.externalResources">
<code class="descname">externalResources</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.externalResources" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.indices">
<code class="descname">indices</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.indices" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.metaType">
<code class="descname">metaType</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.metaType" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.pbi">
<code class="descname">pbi</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.pbi" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.reference">
<code class="descname">reference</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.reference" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.resourceId">
<code class="descname">resourceId</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.resourceId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.scraps">
<code class="descname">scraps</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.scraps" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.tags">
<code class="descname">tags</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.tags" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResource.timeStampedName">
<code class="descname">timeStampedName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResource.timeStampedName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ExternalResources</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.addResources">
<code class="descname">addResources</code><span class="sig-paren">(</span><em>resourceIds</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.addResources" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new external reference with the given uris. If you&#8217;re looking
to add ExternalResource objects, append() or extend() them instead.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd>resourceIds: a list of uris as strings</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.resourceIds">
<code class="descname">resourceIds</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.resourceIds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.resources">
<code class="descname">resources</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.resources" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.ExternalResources.sort">
<code class="descname">sort</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ExternalResources.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>In theory we could sort the ExternalResource objects, but that
would require opening them</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndex">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">FileIndex</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndex.metaType">
<code class="descname">metaType</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndex.metaType" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndex.resourceId">
<code class="descname">resourceId</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndex.resourceId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndex.timeStampedName">
<code class="descname">timeStampedName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndex.timeStampedName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndices">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">FileIndices</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndices" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndices.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndices.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.FileIndices.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.FileIndices.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Filter">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Filter</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.addRequirement">
<code class="descname">addRequirement</code><span class="sig-paren">(</span><em>name</em>, <em>operator</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.addRequirement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.plist">
<code class="descname">plist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.plist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.pop">
<code class="descname">pop</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.pop" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filter.removeRequirement">
<code class="descname">removeRequirement</code><span class="sig-paren">(</span><em>req</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filter.removeRequirement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Filters">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Filters</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.addRequirement">
<code class="descname">addRequirement</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.addRequirement" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to add requirements. Members of the list will be considered
options for fulfilling this requirement, all other filters will be
duplicated for each option. Use multiple calls to add multiple
requirements to the existing filters. Use removeRequirement first to
not add conflicting filters.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd>name: The name of the requirement, e.g. &#8216;rq&#8217;
options: A list of (operator, value) tuples, e.g. (&#8216;&gt;&#8217;, &#8216;0.85&#8217;)</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.filterIndexRecords">
<code class="descname">filterIndexRecords</code><span class="sig-paren">(</span><em>indexRecords</em>, <em>nameMap</em>, <em>movieMap</em>, <em>readType='bam'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.filterIndexRecords" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.fromString">
<code class="descname">fromString</code><span class="sig-paren">(</span><em>filterString</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.fromString" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.mapRequirement">
<code class="descname">mapRequirement</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.mapRequirement" title="Permalink to this definition">¶</a></dt>
<dd><p>Add requirements to each of the existing requirements, mapped one
to one</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.opMap">
<code class="descname">opMap</code><span class="sig-paren">(</span><em>op</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.opMap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.registerCallback">
<code class="descname">registerCallback</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.registerCallback" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.removeRequirement">
<code class="descname">removeRequirement</code><span class="sig-paren">(</span><em>req</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.removeRequirement" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.testCompatibility">
<code class="descname">testCompatibility</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.testCompatibility" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.testParam">
<code class="descname">testParam</code><span class="sig-paren">(</span><em>param</em>, <em>value</em>, <em>testType=&lt;type 'str'&gt;</em>, <em>oper='='</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.testParam" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Filters.tests">
<code class="descname">tests</code><span class="sig-paren">(</span><em>readType='bam'</em>, <em>tIdMap=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Filters.tests" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.OutputOptions">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">OutputOptions</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.OutputOptions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.OutputOptions.collectionPathUri">
<code class="descname">collectionPathUri</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.OutputOptions.collectionPathUri" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.OutputOptions.copyFiles">
<code class="descname">copyFiles</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.OutputOptions.copyFiles" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.OutputOptions.resultsFolder">
<code class="descname">resultsFolder</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.OutputOptions.resultsFolder" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.ParentTool">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">ParentTool</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.ParentTool" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">PbiFlags</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.ADAPTER_AFTER">
<code class="descname">ADAPTER_AFTER</code><em class="property"> = 2</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.ADAPTER_AFTER" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.ADAPTER_BEFORE">
<code class="descname">ADAPTER_BEFORE</code><em class="property"> = 1</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.ADAPTER_BEFORE" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.BARCODE_AFTER">
<code class="descname">BARCODE_AFTER</code><em class="property"> = 8</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.BARCODE_AFTER" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.BARCODE_BEFORE">
<code class="descname">BARCODE_BEFORE</code><em class="property"> = 4</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.BARCODE_BEFORE" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.FORWARD_PASS">
<code class="descname">FORWARD_PASS</code><em class="property"> = 16</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.FORWARD_PASS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.REVERSE_PASS">
<code class="descname">REVERSE_PASS</code><em class="property"> = 32</em><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.REVERSE_PASS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="classmethod">
<dt id="pbcore.io.dataset.DataSetMembers.PbiFlags.flagMap">
<em class="property">classmethod </em><code class="descname">flagMap</code><span class="sig-paren">(</span><em>flag</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PbiFlags.flagMap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.PrimaryMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">PrimaryMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="docutils">
<dt>Doctest:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">tempfile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">SubreadSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data.datasets</span> <span class="kn">as</span> <span class="nn">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">getXml</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">primary</span><span class="o">.</span><span class="n">outputOptions</span><span class="o">.</span><span class="n">resultsFolder</span>
<span class="go">&#39;Analysis_Results&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">primary</span><span class="o">.</span><span class="n">outputOptions</span><span class="o">.</span><span class="n">resultsFolder</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="s">&#39;BetterAnalysis_Results&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">primary</span><span class="o">.</span><span class="n">outputOptions</span><span class="o">.</span><span class="n">resultsFolder</span>
<span class="go">&#39;BetterAnalysis_Results&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">(</span><span class="n">suffix</span><span class="o">=</span><span class="s">&quot;dataset-doctest&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">outXml</span> <span class="o">=</span> <span class="s">&#39;xml:&#39;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s">&#39;tempfile.xml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outXml</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">SubreadSet</span><span class="p">(</span><span class="n">outXml</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds2</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">primary</span><span class="o">.</span><span class="n">outputOptions</span><span class="o">.</span><span class="n">resultsFolder</span>
<span class="go">&#39;BetterAnalysis_Results&#39;</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PrimaryMetadata.automationName">
<code class="descname">automationName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata.automationName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PrimaryMetadata.configFileName">
<code class="descname">configFileName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata.configFileName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PrimaryMetadata.outputOptions">
<code class="descname">outputOptions</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata.outputOptions" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.PrimaryMetadata.sequencingCondition">
<code class="descname">sequencingCondition</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.PrimaryMetadata.sequencingCondition" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Properties">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Properties</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Properties" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Properties.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Properties.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Properties.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Properties.__iter__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.Properties.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Properties.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Property">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Property</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Property" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Property.name">
<code class="descname">name</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Property.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Property.operator">
<code class="descname">operator</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Property.operator" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Property.value">
<code class="descname">value</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Property.value" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.Provenance">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">Provenance</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Provenance" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The metadata concerning this dataset&#8217;s provenance</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Provenance.createdBy">
<code class="descname">createdBy</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Provenance.createdBy" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.Provenance.parentTool">
<code class="descname">parentTool</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.Provenance.parentTool" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">RecordWrapper</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>The base functionality of a metadata element.</p>
<p>Many of the methods here are intended for use with children of
RecordWrapper (e.g. append, extend). Methods in child classes often provide
similar functionality for more raw inputs (e.g. resourceIds as strings)</p>
<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.__getitem__">
<code class="descname">__getitem__</code><span class="sig-paren">(</span><em>tag</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Try to get the a specific child (only useful in simple cases where
children will not be wrapped in a special wrapper object, returns the
first instance of &#8216;tag&#8217;)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.__iter__">
<code class="descname">__iter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Get each child iteratively (only useful in simple cases where
children will not be wrapped in a special wrapper object)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.__repr__">
<code class="descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a pretty string represenation of this object:</p>
<p>&#8220;&lt;type tag text attribs children&gt;&#8221;</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.addMetadata">
<code class="descname">addMetadata</code><span class="sig-paren">(</span><em>key</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.addMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a key, value pair to this metadata object (attributes)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.append">
<code class="descname">append</code><span class="sig-paren">(</span><em>newMember</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append to the actual list of child elements</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.attrib">
<code class="descname">attrib</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.attrib" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.description">
<code class="descname">description</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.description" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.extend">
<code class="descname">extend</code><span class="sig-paren">(</span><em>newMembers</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend the actual list of child elements</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.findChildren">
<code class="descname">findChildren</code><span class="sig-paren">(</span><em>tag</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.findChildren" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.getMemberV">
<code class="descname">getMemberV</code><span class="sig-paren">(</span><em>tag</em>, <em>container='text'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.getMemberV" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic accessor for the contents of the children of this element,
without having to interface with them directly</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.getV">
<code class="descname">getV</code><span class="sig-paren">(</span><em>container='text'</em>, <em>tag=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.getV" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic accessor for the contents of this element&#8217;s &#8216;attrib&#8217; or
&#8216;text&#8217; fields</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.index">
<code class="descname">index</code><span class="sig-paren">(</span><em>tag</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the index in &#8216;children&#8217; list of item with &#8216;tag&#8217; member</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.metadata">
<code class="descname">metadata</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleaner accessor for this node&#8217;s attributes. Returns mutable,
doesn&#8217;t need setter</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.metaname">
<code class="descname">metaname</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.metaname" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleaner accessor for this node&#8217;s tag</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.metavalue">
<code class="descname">metavalue</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.metavalue" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleaner accessor for this node&#8217;s text</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.name">
<code class="descname">name</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.namespace">
<code class="descname">namespace</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.namespace" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.pop">
<code class="descname">pop</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.pop" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.pruneChildrenTo">
<code class="descname">pruneChildrenTo</code><span class="sig-paren">(</span><em>whitelist</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.pruneChildrenTo" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.removeChildren">
<code class="descname">removeChildren</code><span class="sig-paren">(</span><em>tag</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.removeChildren" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.setMemberV">
<code class="descname">setMemberV</code><span class="sig-paren">(</span><em>tag</em>, <em>value</em>, <em>container='text'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.setMemberV" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic accessor for the contents of the children of this element,
without having to interface with them directly</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.setV">
<code class="descname">setV</code><span class="sig-paren">(</span><em>value</em>, <em>container='text'</em>, <em>tag=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.setV" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic accessor for the contents of this element&#8217;s &#8216;attrib&#8217; or
&#8216;text&#8217; fields</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.submetadata">
<code class="descname">submetadata</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.submetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleaner accessor for wrapped versions of this node&#8217;s children.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.subrecords">
<code class="descname">subrecords</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.subrecords" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleaner accessor for this node&#8217;s children. Returns mutable, doesn&#8217;t
need setter</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.tags">
<code class="descname">tags</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.tags" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the list of tags for children in this element</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.value">
<code class="descname">value</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.value" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RecordWrapper.version">
<code class="descname">version</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper.version" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.RunDetailsMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">RunDetailsMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RunDetailsMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RunDetailsMetadata.name">
<code class="descname">name</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RunDetailsMetadata.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.RunDetailsMetadata.timeStampedName">
<code class="descname">timeStampedName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.RunDetailsMetadata.timeStampedName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">StatsMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<p>The metadata from the machine sts.xml</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.adapterDimerFraction">
<code class="descname">adapterDimerFraction</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.adapterDimerFraction" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.controlReadLenDist">
<code class="descname">controlReadLenDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.controlReadLenDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.controlReadQualDist">
<code class="descname">controlReadQualDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.controlReadQualDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.insertReadLenDist">
<code class="descname">insertReadLenDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.insertReadLenDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.insertReadQualDist">
<code class="descname">insertReadQualDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.insertReadQualDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.medianInsertDist">
<code class="descname">medianInsertDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.medianInsertDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.medianInsertDists">
<code class="descname">medianInsertDists</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.medianInsertDists" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.numSequencingZmws">
<code class="descname">numSequencingZmws</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.numSequencingZmws" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.prodDist">
<code class="descname">prodDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.prodDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.readLenDist">
<code class="descname">readLenDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.readLenDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.readLenDists">
<code class="descname">readLenDists</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.readLenDists" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.readQualDist">
<code class="descname">readQualDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.readQualDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.readQualDists">
<code class="descname">readQualDists</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.readQualDists" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.readTypeDist">
<code class="descname">readTypeDist</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.readTypeDist" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.StatsMetadata.shortInsertFraction">
<code class="descname">shortInsertFraction</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.StatsMetadata.shortInsertFraction" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.SubreadSetMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">SubreadSetMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.SubreadSetMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.DataSetMetadata" title="pbcore.io.dataset.DataSetMembers.DataSetMetadata"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.DataSetMetadata</span></code></a></p>
<p>The DataSetMetadata subtype specific to SubreadSets. Deals explicitly
with the merging of Collections and BioSamples metadata hierarchies.</p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.bioSamples">
<code class="descname">bioSamples</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.bioSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of wrappers around BioSamples elements of the Metadata
Record</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.collections">
<code class="descname">collections</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.collections" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of wrappers around Collections elements of the
Metadata Record</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.SubreadSetMetadata.merge" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata">
<em class="property">class </em><code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">WellSampleMetadata</code><span class="sig-paren">(</span><em>record=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pbcore.io.dataset.DataSetMembers.RecordWrapper" title="pbcore.io.dataset.DataSetMembers.RecordWrapper"><code class="xref py py-class docutils literal"><span class="pre">pbcore.io.dataset.DataSetMembers.RecordWrapper</span></code></a></p>
<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.bioSamplePointers">
<code class="descname">bioSamplePointers</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.bioSamplePointers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.comments">
<code class="descname">comments</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.comments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.concentration">
<code class="descname">concentration</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.concentration" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.sampleReuseEnabled">
<code class="descname">sampleReuseEnabled</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.sampleReuseEnabled" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.sizeSelectionEnabled">
<code class="descname">sizeSelectionEnabled</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.sizeSelectionEnabled" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.stageHotstartEnabled">
<code class="descname">stageHotstartEnabled</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.stageHotstartEnabled" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.uniqueId">
<code class="descname">uniqueId</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.uniqueId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.useCount">
<code class="descname">useCount</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.useCount" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.dataset.DataSetMembers.WellSampleMetadata.wellName">
<code class="descname">wellName</code><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.WellSampleMetadata.wellName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetMembers.filter_read">
<code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">filter_read</code><span class="sig-paren">(</span><em>accessor</em>, <em>operator</em>, <em>value</em>, <em>read</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.filter_read" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetMembers.getTime">
<code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">getTime</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.getTime" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetMembers.getTimeStampedName">
<code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">getTimeStampedName</code><span class="sig-paren">(</span><em>mType</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.getTimeStampedName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pbcore.io.dataset.DataSetMembers.newUuid">
<code class="descclassname">pbcore.io.dataset.DataSetMembers.</code><code class="descname">newUuid</code><span class="sig-paren">(</span><em>record</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.dataset.DataSetMembers.newUuid" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pbcore.io.dataset</a></li>
<li><a class="reference internal" href="#console-entry-point-usage">Console Entry Point Usage</a></li>
<li><a class="reference internal" href="#usage-examples">Usage Examples</a><ul>
<li><a class="reference internal" href="#filter-reads-cli-version">Filter Reads (CLI version)</a></li>
<li><a class="reference internal" href="#filter-reads-api-version">Filter Reads (API version)</a></li>
<li><a class="reference internal" href="#resequencing-pipeline-cli-version">Resequencing Pipeline (CLI version)</a><ul>
<li><a class="reference internal" href="#tangential-information">Tangential Information</a></li>
</ul>
</li>
<li><a class="reference internal" href="#resequencing-pipeline-api-version">Resequencing Pipeline (API version)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#api-overview">API overview</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/pbcoretools.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pbcoretools 0.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011-2015, Pacific Biosciences.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>